File: sdk/java/src/main/java/com/xxxxteam/tempmail/providers/DisposablemailCom.java
package com.xxxxteam.tempmail.providers;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.xxxxteam.tempmail.Email;
import com.xxxxteam.tempmail.EmailInfo;
import com.xxxxteam.tempmail.HttpResult;
import com.xxxxteam.tempmail.HttpClient;
import com.xxxxteam.tempmail.Json;
import com.xxxxteam.tempmail.Normalizer;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * disposablemail.com 渠道 — https://www.disposablemail.com
 *
 * <p>GET / 获取 session cookie + CSRF → GET /index/index 创建邮箱
 * → GET /index/refresh 获取邮件列表 → GET /email/id/{id} 获取正文。</p>
 */
public final class DisposablemailCom {

    private static final String BASE_URL = "https://www.disposablemail.com";
    private static final String TOK_PREFIX = "dmc1:";
    private static final Pattern CSRF_RE = Pattern.compile("CSRF=\"([^\"]+)\"");

    private DisposablemailCom() {
    }

    private static Map<String, String> ajaxHeaders(String cookie) {
        Map<String, String> h = new LinkedHashMap<>();
        h.put("User-Agent", "Mozilla/5.0");
        h.put("Accept", "application/json, text/javascript, */*; q=0.01");
        h.put("X-Requested-With", "XMLHttpRequest");
        h.put("Referer", BASE_URL + "/");
        if (cookie != null && !cookie.isEmpty()) h.put("Cookie", cookie);
        return h;
    }

    private static String mergeCookies(String prev, HttpResult resp) {
        Map<String, String> jar = new LinkedHashMap<>();
        if (prev != null) {
            for (String part : prev.split(";")) {
                String p = part.trim();
                int eq = p.indexOf('=');
                if (eq > 0) jar.put(p.substring(0, eq).trim(), p.substring(eq + 1).trim());
            }
        }
        for (String sc : resp.getSetCookies()) {
            String seg = sc.split(";")[0].trim();
            int eq = seg.indexOf('=');
            if (eq > 0) jar.put(seg.substring(0, eq), seg.substring(eq + 1));
        }
        StringBuilder sb = new StringBuilder();
        for (var e : jar.entrySet()) {
            if (sb.length() > 0) sb.append("; ");
            sb.append(e.getKey()).append("=").append(e.getValue());
        }
        return sb.toString();
    }

    private static String encodeToken(String csrf, String cookie) {
        String raw = Json.serialize(Map.of("t", csrf, "c", cookie));
        return TOK_PREFIX + Base64.getEncoder().encodeToString(raw.getBytes(StandardCharsets.UTF_8));
    }

    private static String[] decodeToken(String token) {
        if (token == null || !token.startsWith(TOK_PREFIX)) {
            throw new RuntimeException("disposablemail-com: 无效的 token");
        }
        String raw = new String(Base64.getDecoder().decode(token.substring(TOK_PREFIX.length())), StandardCharsets.UTF_8);
        JsonObject obj = Json.parseObject(raw);
        if (obj == null) throw new RuntimeException("disposablemail-com: 无效的 token");
        return new String[]{Json.str(obj, "t"), Json.str(obj, "c")};
    }

    /**
     * 创建 disposablemail.com 临时邮箱。
     *
     * @return 邮箱信息
     */
    public static EmailInfo generate() {
        Map<String, String> bh = new LinkedHashMap<>();
        bh.put("User-Agent", "Mozilla/5.0");
        bh.put("Accept", "text/html");
        HttpResult r1 = HttpClient.get(BASE_URL + "/", bh);
        r1.ensureSuccess();
        String cookie = mergeCookies("", r1);

        Matcher m = CSRF_RE.matcher(r1.getBody());
        if (!m.find()) throw new RuntimeException("disposablemail-com: 未能提取 CSRF token");
        String csrf = m.group(1);

        HttpResult r2 = HttpClient.get(
                BASE_URL + "/index/index?csrf_token=" + ProviderUtil.urlEncode(csrf),
                ajaxHeaders(cookie));
        r2.ensureSuccess();
        cookie = mergeCookies(cookie, r2);

        JsonObject data = Json.parseObject(r2.getBody());
        if (data == null) throw new RuntimeException("disposablemail-com: 创建邮箱响应格式异常");
        String email = Json.str(data, "email").trim();
        if (email.isEmpty() || !email.contains("@")) {
            throw new RuntimeException("disposablemail-com: 邮箱地址无效");
        }

        return new EmailInfo("disposablemail-com", email, encodeToken(csrf, cookie), null, null);
    }

    /**
     * 获取邮件列表。
     *
     * @param email 邮箱地址
     * @param token 编码的 token
     * @return 邮件列表
     */
    public static List<Email> getEmails(String email, String token) {
        String[] parts = decodeToken(token);
        String cookie = parts[1];

        HttpResult resp = HttpClient.get(BASE_URL + "/index/refresh", ajaxHeaders(cookie));
        resp.ensureSuccess();

        String trimmed = resp.getBody().trim();
        if ("0".equals(trimmed) || trimmed.isEmpty() || "[]".equals(trimmed)) {
            return new ArrayList<>();
        }

        JsonElement parsed = Json.parse(resp.getBody());
        if (parsed == null || !parsed.isJsonArray()) return new ArrayList<>();

        List<Email> out = new ArrayList<>();
        for (JsonElement item : parsed.getAsJsonArray()) {
            if (!item.isJsonObject()) continue;
            JsonObject msg = item.getAsJsonObject();
            String mailId = Json.str(msg, "id");
            String htmlBody = fetchBody(cookie, mailId);

            Map<String, Object> row = new LinkedHashMap<>();
            row.put("id", mailId);
            row.put("from", Json.str(msg, "od"));
            row.put("to", email);
            row.put("subject", Json.str(msg, "predmet"));
            row.put("html", htmlBody);
            row.put("date", Json.str(msg, "kdy"));
            out.add(Normalizer.normalizeEmail(row, email));
        }
        return out;
    }

    private static String fetchBody(String cookie, String mailId) {
        if (mailId == null || mailId.isEmpty()) return "";
        try {
            HttpResult resp = HttpClient.get(BASE_URL + "/email/id/" + mailId, ajaxHeaders(cookie));
            return resp.isOk() ? resp.getBody() : "";
        } catch (RuntimeException e) {
            return "";
        }
    }
}

File: sdk/kotlin/src/main/kotlin/com/xxxxteam/tempmail/providers/DisposablemailCom.kt
package com.xxxxteam.tempmail.providers

import com.xxxxteam.tempmail.Email
import com.xxxxteam.tempmail.EmailInfo
import com.xxxxteam.tempmail.Http
import com.xxxxteam.tempmail.Normalize
import com.xxxxteam.tempmail.Provider
import kotlinx.serialization.json.*
import java.nio.charset.StandardCharsets
import java.util.Base64

/**
 * disposablemail.com 渠道（www.disposablemail.com）。
 *
 * GET / 取 session cookie + CSRF，GET /index/index 创建邮箱，
 * GET /index/refresh 收信，GET /email/id/{id} 取正文。token base64 编码，前缀 "dmc1:"。
 */
object DisposablemailCom : Provider {

    private const val BASE_URL = "https://www.disposablemail.com"
    private const val TOK_PREFIX = "dmc1:"
    private val CSRF_RE = Regex("CSRF=\"([^\"]+)\"")

    private fun ajaxHeaders(cookie: String): Map<String, String> {
        val h = mutableMapOf(
            "User-Agent" to "Mozilla/5.0",
            "Accept" to "application/json, text/javascript, */*; q=0.01",
            "X-Requested-With" to "XMLHttpRequest",
            "Referer" to "$BASE_URL/",
        )
        if (cookie.isNotEmpty()) h["Cookie"] = cookie
        return h
    }

    /** 创建临时邮箱。 */
    override suspend fun generate(): EmailInfo {
        val r1 = ProviderUtil.httpGet("$BASE_URL/",
            mapOf("User-Agent" to "Mozilla/5.0", "Accept" to "text/html"))
        r1.ensureSuccess()
        var cookie = mergeCookies("", r1)
        val csrf = CSRF_RE.find(r1.body)?.groupValues?.get(1)
            ?: throw RuntimeException("disposablemail-com: 未能提取 CSRF token")
        val r2 = ProviderUtil.httpGet(
            "$BASE_URL/index/index?csrf_token=${ProviderUtil.urlEncode(csrf)}", ajaxHeaders(cookie))
        r2.ensureSuccess()
        cookie = mergeCookies(cookie, r2)
        val data = Http.json.parseToJsonElement(r2.body) as? JsonObject
            ?: throw RuntimeException("disposablemail-com: 创建邮箱响应格式异常")
        val email = jsonStr(data, "email").trim()
        if (email.isBlank() || !email.contains("@")) throw RuntimeException("disposablemail-com: 邮箱地址无效")
        return EmailInfo(email = email, channel = "disposablemail-com", token = encodeToken(csrf, cookie))
    }

    /** 获取邮件列表。 */
    override suspend fun getEmails(info: EmailInfo): List<Email> {
        val email = info.email.trim()
        val cookie = decodeToken(info.token)
        val resp = ProviderUtil.httpGet("$BASE_URL/index/refresh", ajaxHeaders(cookie))
        resp.ensureSuccess()
        val trimmed = resp.body.trim()
        if (trimmed == "0" || trimmed.isEmpty() || trimmed == "[]") return emptyList()
        val arr = Http.json.parseToJsonElement(resp.body) as? JsonArray ?: return emptyList()
        return arr.mapNotNull { it as? JsonObject }.map { msg ->
            val mailId = jsonStr(msg, "id")
            val flat = mapOf(
                "id" to mailId,
                "from" to jsonStr(msg, "od"),
                "to" to email,
                "subject" to jsonStr(msg, "predmet"),
                "html" to fetchBody(mailId, cookie),
                "date" to jsonStr(msg, "kdy"),
            )
            Normalize.fromMap(flat, email)
        }
    }

    /** 拉取单封邮件正文，失败返回空串。 */
    private suspend fun fetchBody(mailId: String, cookie: String): String {
        if (mailId.isEmpty()) return ""
        return try {
            val resp = ProviderUtil.httpGet("$BASE_URL/email/id/$mailId", ajaxHeaders(cookie))
            if (resp.isOk) resp.body else ""
        } catch (_: Exception) {
            ""
        }
    }

    /** 编码 token（base64，前缀 dmc1:），存储 {t:csrf, c:cookie}。 */
    private fun encodeToken(csrf: String, cookie: String): String {
        val raw = buildJsonObject { put("t", csrf); put("c", cookie) }.toString()
        return TOK_PREFIX + Base64.getEncoder().encodeToString(raw.toByteArray(StandardCharsets.UTF_8))
    }

    /** 解码 token，返回 cookie 串。 */
    private fun decodeToken(token: String): String {
        if (!token.startsWith(TOK_PREFIX)) throw RuntimeException("disposablemail-com: 无效的 token")
        val raw = String(Base64.getDecoder().decode(token.substring(TOK_PREFIX.length)), StandardCharsets.UTF_8)
        val obj = Http.json.parseToJsonElement(raw) as? JsonObject
            ?: throw RuntimeException("disposablemail-com: 无效的 token")
        return jsonStr(obj, "c")
    }

    /** 合并 cookie 串与响应新 cookie。 */
    private fun mergeCookies(prev: String, resp: HttpResp): String {
        val jar = LinkedHashMap<String, String>()
        prev.split(";").forEach { part ->
            val p = part.trim()
            val eq = p.indexOf('=')
            if (eq > 0) jar[p.substring(0, eq).trim()] = p.substring(eq + 1).trim()
        }
        resp.setCookies.forEach { sc ->
            val seg = sc.split(";")[0].trim()
            val eq = seg.indexOf('=')
            if (eq > 0) jar[seg.substring(0, eq)] = seg.substring(eq + 1)
        }
        return jar.entries.joinToString("; ") { "${it.key}=${it.value}" }
    }
}

/** 从 JSON 对象读取字符串字段，缺失返回空串（文件私有）。 */
private fun jsonStr(obj: JsonObject, key: String): String =
    (obj[key] as? JsonPrimitive)?.contentOrNull ?: ""
